File: BoardGames/Board.cpp
#include <iomanip>
#include <iostream>
#include "Board.hpp"
#include "Utils.hpp"


Board::Board(int row, int column) : rows(row), columns(column) {

	grid.resize(rows);
	for (int i = 0; i < rows; ++i) {
		grid[i].resize(columns);
		for (int j = 0; j < columns; ++j) {
			grid[i][j] = std::make_shared<Cell>(0);
		}
	}
}


std::ostream& operator<<(std::ostream& os, const Board& board) {
	for (int r = 0; r < board.rows; r++) {
		for (int c = 0; c < board.columns; c++) {
			os << std::setw(6) << board.grid[r][c]->GetValue();
		}
		os << "\n";
	}
	return os;
}


int Board::GetValue(int row, int col) {
	if (row < 0 || row > rows) return 0;
	else if (col < 0 || col > columns) return 0;
	else return grid[row][col]->GetValue();
}


void Board::SetValue(int row, int col, int val) {
	if (row >= 0 || row < rows)
	if (col >= 0 || col < columns)
	grid[row][col]->SetValue(val);
}


int Board::PointSum(int col) {
	if (col < 0 || col >= columns) return 0;

	std::vector<int> values;
	for (int i = 0; i < rows - 1; i++) {
		values.push_back(grid[i][col]->GetValue());
	}

	int pointSum = SumScores(values);

	grid[rows-1][col]->SetValue(pointSum);
	return pointSum;
}

std::shared_ptr<Board> Board::ShallowClone() const {
	auto clone = std::make_shared<Board>(rows, columns);
	clone->grid = this->grid;
	return clone;
}


std::shared_ptr<Board> Board::DeepClone() const {
	auto clone = std::make_shared<Board>(rows, columns);

	for (int i = 0; i < rows; ++i) {
		for (int j = 0; j < columns; ++j) {
			clone->grid[i][j] = std::make_shared<Cell>(*grid[i][j]);
		}
	}

	return clone;
}

File: BoardGamesSharp/Core/Board.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace BoardGamesSharp.Core;

public class Board : ICloneable
{
    private int rows;
    private int columns;
    private Cell[,] grid;

    public Board(int row, int column) 
    {  
        rows = row; 
        columns = column; 
        grid = new Cell[rows, columns];

        for (int r = 0; r < rows; r++)
        {
            for (int c = 0; c < columns; c++)
            {
                grid[r, c] = new Cell();
            }
        }
    }


    public void Show()
    {
        for(int r = 0; r < rows; r++)
        {
            for (int c = 0; c < columns ; c++)
            {
                Console.Write($"{grid[r, c].val,6}");
            }
            Console.WriteLine();
        }
    }

    public int GetValue(int row, int column)
    {
        if (row < 0 || row >= rows) return 0;
        else if (column < 0 || column >= columns) return 0;
        else return grid[row, column].val;
    }


    public void SetValue(int row, int column, int val)
    {
        if (row >= 0 && row < rows)
        if (column >= 0 && column < columns)
        grid[row, column].val = val;
    }


    public int PointSum(int column)
    {
        if (column < 0 || column >= columns) return 0;

        var values = new List<int>();

        for (int i = 0; i < rows - 1; i++)
        {
            values.Add(grid[i, column].val);
        }

        int pointSum = Utils.SumScores(values);

        grid[rows-1, column].val = pointSum;
        return pointSum;
    }

    public object Clone()
    {
        return (Board)this.MemberwiseClone();
    }

    public object DeepClone()
    {
        return new Board(rows, columns);
    }
}


